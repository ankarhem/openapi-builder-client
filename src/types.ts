import { ConditionalExcept, Get, Simplify, ValueOf } from 'type-fest';
import { paths } from '../openapi/petstore';

type OmitNever<T> = ConditionalExcept<T, never>;

export type GetPaths<OpenAPIPaths> = OmitNever<{
  [Path in keyof OpenAPIPaths]: Extract<
    OpenAPIPaths[Path],
    { get: any }
  >['get'];
}>;
export type PostPaths<OpenAPIPaths> = OmitNever<{
  [Path in keyof OpenAPIPaths]: Extract<
    OpenAPIPaths[Path],
    { post: any }
  >['post'];
}>;
export type PutPaths<OpenAPIPaths> = OmitNever<{
  [Path in keyof OpenAPIPaths]: Extract<
    OpenAPIPaths[Path],
    { put: any }
  >['put'];
}>;
export type DeletePaths<OpenAPIPaths> = OmitNever<{
  [Path in keyof OpenAPIPaths]: Extract<
    OpenAPIPaths[Path],
    { delete: any }
  >['delete'];
}>;
export type PatchPaths<OpenAPIPaths> = OmitNever<{
  [Path in keyof OpenAPIPaths]: Extract<
    OpenAPIPaths[Path],
    { patch: any }
  >['patch'];
}>;

export type ResponseOf<Path> = Simplify<
  {
    [Code in keyof Get<Path, 'responses'>]: Omit<
      Response,
      'status' | 'json'
    > & {
      status: Code;
      json(): Promise<ValueOf<Get<Get<Path, 'responses'>[Code], 'content'>>>;
    };
  }[keyof Get<Path, 'responses'>]
>;

export type QueryParamsOf<Path> = Get<Path, 'parameters.query'>;
export type PathParamsOf<Path> = Get<Path, 'parameters.path'>;
export type BodyOf<Path> = ValueOf<Get<Path, 'requestBody.content'>>;

export type TestPathsGet = GetPaths<paths>;
type TestResponse = ResponseOf<TestPathsGet['/pet/findByTags']>;

export type TestPathsPost = PostPaths<paths>;
type TestPathParams = PathParamsOf<TestPathsPost['/pet/{petId}']>;
type TestQueryParams = PathParamsOf<TestPathsPost['/pet/{petId}']>;
type TestBody = BodyOf<TestPathsPost['/pet/{petId}']>;

async function testing() {
  const r = '' as unknown as TestResponse;

  if (r.status === 400) {
    const data = await r.json();
  }

  const data = await r.json();
}
