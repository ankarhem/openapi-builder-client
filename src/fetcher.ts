import { Fetcher, MiddlewareFunction } from './types';

export class OwnedFetcher {
  private sender: Fetcher;
  constructor(fetcher: Fetcher) {
    this.sender = fetcher;
  }

  withRetries(retries: number = 0) {
    if (retries <= 0) return this;

    const fetchWithRetries = async (
      url: string,
      init: RequestInit,
      retryCount: number = 0
    ): ReturnType<Fetcher> => {
      try {
        return await this.send(url, init);
      } catch (error) {
        if (retryCount < (retries || 0)) {
          return await fetchWithRetries(url, init, retryCount + 1);
        }
        throw error;
      }
    };

    return new OwnedFetcher(fetchWithRetries);
  }

  withMiddlewares(middlewares: MiddlewareFunction[] = []) {
    if (middlewares.length === 0) return this;

    const fetchWithMiddlewares = (
      url: string,
      init: RequestInit,
      index: number = 0
    ): ReturnType<Fetcher> => {
      if (!middlewares || index === middlewares.length) {
        return this.sender(url, init);
      }
      const current = middlewares?.[index];
      return current(url, init, (nextUrl, nextInit) =>
        fetchWithMiddlewares(nextUrl, nextInit, index + 1)
      );
    };

    return new OwnedFetcher(fetchWithMiddlewares);
  }

  send(url: string, init: RequestInit) {
    return this.sender(url, init);
  }
}
