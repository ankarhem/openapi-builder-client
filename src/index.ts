export { Client } from './client';
export { defaultFormatter, joinFormatter } from './search';
import { Fetcher, ClientOptions, MiddlewareFunction } from './types';
export type { Fetcher, ClientOptions, MiddlewareFunction };
