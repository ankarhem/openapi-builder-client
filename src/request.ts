import {
  JsonBodyOf,
  HeaderParamsOf,
  NextOwnedRequest,
  OwnedRequestState,
  PathParamsOf,
  QueryParamsOf,
  ResponseOf,
  FormDataOf,
} from './types';

type BodyMethods = 'body' | 'form';

export class OwnedRequest<Path, UsedMethods extends string = ''> {
  private state: OwnedRequestState = {
    path: {},
    query: {},
    headers: {},
    body: undefined,
    extras: undefined,
  };

  private _send: (state: OwnedRequestState) => Promise<ResponseOf<Path>>;

  constructor(send: (state: OwnedRequestState) => Promise<ResponseOf<Path>>) {
    this._send = send;
    return this;
  }

  __withDynamicTyping() {
    return this as NextOwnedRequest<Path, UsedMethods>;
  }

  path(
    params: PathParamsOf<Path>
  ): NextOwnedRequest<Path, UsedMethods | 'path'> {
    if (params) {
      this.state.path = params;
    }
    return this as NextOwnedRequest<Path, UsedMethods | 'path'>;
  }
  __path(params: Record<string, string | number>) {
    return this.path(params as any);
  }

  query(params: QueryParamsOf<Path>) {
    if (params) {
      this.state.query = params;
    }
    return this as NextOwnedRequest<Path, UsedMethods | 'query'>;
  }
  __query(params: Record<string, any>) {
    return this.query(params as any);
  }

  body(body: JsonBodyOf<Path>) {
    if (body) {
      this.state.body = body;
      this.state.headers['Content-Type'] = 'application/json';
    }
    return this as NextOwnedRequest<Path, UsedMethods | BodyMethods>;
  }
  __body(body: Record<string, any>) {
    return this.body(body as any);
  }

  form(data: FormDataOf<Path>) {
    if (data) {
      this.state.body = data;
      this.state.headers['Content-Type'] = 'application/x-www-form-urlencoded';
    }
    return this as NextOwnedRequest<Path, UsedMethods | BodyMethods>;
  }
  __form(data: Record<string, any>) {
    return this.form(data as any);
  }

  headers(headers: HeaderParamsOf<Path>) {
    if (headers) {
      this.state.headers = {
        ...this.state.headers,
        ...headers,
      };
    }
    return this as NextOwnedRequest<Path, UsedMethods | 'headers'>;
  }
  __headers(headers: Record<string, string>) {
    return this.headers(headers as any);
  }

  send(extras?: Pick<RequestInit, 'signal'>) {
    this.state.extras = extras;
    return this._send(this.state);
  }
}
