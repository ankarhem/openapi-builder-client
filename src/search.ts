import { Primitive } from 'type-fest';
import { FormBodyFormatter } from './types';

/**
 * Repeats key for array values.
 * Stringifies objects.
 */
export const defaultFormatter: FormBodyFormatter = (data) => {
  const formData = new FormData();
  for (const [key, value] of Object.entries(data)) {
    if (typeof value === 'undefined') continue;

    if (Array.isArray(value)) {
      value
        .map((item) => {
          if (typeof item === 'object') {
            return JSON.stringify(item);
          }
          return item.toString();
        })
        .forEach((s) => formData.append(key, s));
      continue;
    }

    if (typeof value === 'object') {
      formData.append(key, JSON.stringify(value));
      continue;
    }

    if (value instanceof File) {
      formData.append(key, value);
    }

    formData.append(key, value.toString());
  }

  return formData;
};

/**
 * Joins array values with `,`
 * Stringifies objects.
 */
export const joinFormatter: FormBodyFormatter = (data) => {
  const formData = new FormData();
  for (const [key, value] of Object.entries(data)) {
    if (typeof value === 'undefined') continue;

    if (Array.isArray(value)) {
      const valueString = value
        .map((item) => {
          if (typeof item === 'object') {
            return JSON.stringify(item);
          }
          return item.toString();
        })
        .join(',');
      formData.append(key, valueString);
      continue;
    }

    if (typeof value === 'object') {
      formData.append(key, JSON.stringify(value));
      continue;
    }

    if (value instanceof File) {
      formData.append(key, value);
    }

    formData.append(key, value.toString());
  }

  return formData;
};

class PathFormatter {
  static flatten(
    data: Record<string, any> | Array<any>,
    prefix = ''
  ): Record<string, string | File> {
    if (!data) return {};

    if (!Array.isArray(data) && prefix.at(-1) === ']') {
      prefix += '.';
    }

    let flattenedObject: Record<string, string | File> = {};
    for (const [_key, value] of Object.entries(data)) {
      const key = prefix + _key;
      if (typeof value === 'undefined') continue;

      if (Array.isArray(value)) {
        value.forEach((item, idx) => {
          if (typeof item === 'object') {
            flattenedObject = {
              ...flattenedObject,
              ...PathFormatter.flatten(item, `${key}[${idx}]`),
            };
            return;
          }

          flattenedObject[`${key}[${idx}]`] = item.toString();
        });
        continue;
      }

      if (typeof value === 'object') {
        flattenedObject = {
          ...flattenedObject,
          ...PathFormatter.flatten(value, `${key}.`),
        };
        continue;
      }

      if (value instanceof File) {
        flattenedObject[key] = value;
        continue;
      }

      flattenedObject[key] = value.toString();
    }
    return flattenedObject;
  }

  static format(data: Record<string, any>) {
    const formData = new FormData();
    Object.entries(PathFormatter.flatten(data)).forEach(([key, value]) => {
      formData.append(key, value);
    });
    return formData;
  }
}

/**
 * Converts object with key as path to value
 * eg.
 * ```json
 * {
 *   "names": [
 *     {
 *       "firstName": "hello"
 *     }
 *   ]
 * }
 * ```
 * `?names[0].firstName=hello`
 */
export const pathFormatter: FormBodyFormatter = PathFormatter.format;
