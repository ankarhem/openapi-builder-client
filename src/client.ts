import { OwnedRequest } from './request';
import {
  ClientOptions,
  DeletePaths,
  Fetcher,
  GetPaths,
  Method,
  OwnedRequestState,
  PatchPaths,
  PostPaths,
  PutPaths,
} from './types';

export class Client<OpenAPIPaths> {
  private options: ClientOptions;
  private wrappedFetch: (url: string, init: RequestInit) => ReturnType<Fetcher>;
  constructor(options: ClientOptions) {
    this.options = options;
    this.wrappedFetch = this.createWrappedFetch();
  }

  /**
   *  Override default ClientOptions configuration
   * @param options {ClientOptions}
   */
  with(options: Partial<ClientOptions>) {
    return new Client<OpenAPIPaths>({
      ...this.options,
      ...options,
      headers: {
        ...this.options.headers,
        ...options.headers,
      },
      middlewares: [
        ...(this.options.middlewares || []),
        ...(options.middlewares || []),
      ],
    });
  }

  private createWrappedFetch(): (
    url: string,
    init: RequestInit
  ) => Promise<Response> {
    if (!this.options.middlewares) return this.options.fetcher;

    const fetchWithRetries = (
      url: string,
      init: RequestInit,
      retries: number = 0
    ): ReturnType<Fetcher> => {
      try {
        return this.options.fetcher(url, init);
      } catch (error) {
        if (retries < (this.options.retries || 0)) {
          return fetchWithRetries(url, init, retries + 1);
        }
        throw error;
      }
    };

    const middlewareHandler = async (
      url: string,
      init: RequestInit,
      index: number = 0
    ): ReturnType<Fetcher> => {
      if (
        !this.options.middlewares ||
        index === this.options.middlewares.length
      ) {
        return fetchWithRetries(url, init);
      }
      const current = this.options.middlewares?.[index];
      return await current(url, init, (nextUrl, nextInit) =>
        middlewareHandler(nextUrl, nextInit!, index + 1)
      );
    };

    return middlewareHandler;
  }

  private send<Path extends keyof OpenAPIPaths>(
    method: Method,
    _path: Path,
    state: OwnedRequestState
  ) {
    let path = _path as string;
    for (const [key, value] of Object.entries(state._pathParams)) {
      path = path.replace(`{${key}}`, value.toString());
    }

    const searchParams = new URLSearchParams();
    for (const [key, value] of Object.entries(state._queryParams)) {
      if (Array.isArray(value)) {
        for (let item in value) {
          searchParams.append(key, item);
        }
        continue;
      }

      searchParams.append(key, value);
    }
    if (searchParams.size > 0) {
      path += `?${searchParams.toString()}`;
    }

    const baseUrl = new URL(this.options.baseUrl);
    const combinedPath = (baseUrl.pathname + path).replace('//', '/');

    const url = new URL(combinedPath, this.options.baseUrl).toString();
    const init: RequestInit = {
      method: method,
      body: Object.keys(state._body).length === 0 ? undefined : state._body,
      headers: {
        ...this.options.headers,
        ...state._headers,
      },
    };

    return this.wrappedFetch(url.toString(), init) as Promise<any>;
  }

  get<PathString extends keyof GetPaths<OpenAPIPaths>>(path: PathString) {
    const ownedRequest = new OwnedRequest<GetPaths<OpenAPIPaths>[PathString]>(
      this.send.bind(this, 'GET', path)
    );
    return ownedRequest.__withDynamicTyping();
  }
  post<PathString extends keyof PostPaths<OpenAPIPaths>>(path: PathString) {
    const ownedRequest = new OwnedRequest<PostPaths<OpenAPIPaths>[PathString]>(
      this.send.bind(this, 'POST', path)
    );
    return ownedRequest.__withDynamicTyping();
  }
  put<PathString extends keyof PutPaths<OpenAPIPaths>>(path: PathString) {
    const ownedRequest = new OwnedRequest<PutPaths<OpenAPIPaths>[PathString]>(
      this.send.bind(this, 'PUT', path)
    );
    return ownedRequest.__withDynamicTyping();
  }
  delete<PathString extends keyof DeletePaths<OpenAPIPaths>>(path: PathString) {
    const ownedRequest = new OwnedRequest<
      DeletePaths<OpenAPIPaths>[PathString]
    >(this.send.bind(this, 'DELETE', path));
    return ownedRequest.__withDynamicTyping();
  }
  patch<PathString extends keyof PatchPaths<OpenAPIPaths>>(path: PathString) {
    const ownedRequest = new OwnedRequest<PatchPaths<OpenAPIPaths>[PathString]>(
      this.send.bind(this, 'PATCH', path)
    );
    return ownedRequest.__withDynamicTyping();
  }
}
