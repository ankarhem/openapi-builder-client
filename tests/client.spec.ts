import { expect, test, describe, Mock, vi } from 'vitest';
import { fetcherWith200Response, mockedClient } from './utils';
import { Fetcher, ConditionFunction } from '../src';
import { joinFormatter, pathFormatter } from '../src/formatters';

describe('Methods', () => {
  const anyClient = mockedClient as any;
  test('GET', async () => {
    anyClient
      .with({
        fetcher: async (url: string, init: RequestInit) => {
          expect(init?.method).toBe('GET');
          return new Response();
        },
      })
      .get('/')
      .send();
  });
  test('POST', async () => {
    anyClient
      .with({
        fetcher: async (url: string, init: RequestInit) => {
          expect(init?.method).toBe('POST');
          return new Response();
        },
      })
      .post('/')
      .send();
  });
  test('PUT', async () => {
    anyClient
      .with({
        fetcher: async (url: string, init: RequestInit) => {
          expect(init?.method).toBe('PUT');
          return new Response();
        },
      })
      .put('/')
      .send();
  });
  test('DELETE', async () => {
    anyClient
      .with({
        fetcher: async (url: string, init: RequestInit) => {
          expect(init?.method).toBe('DELETE');
          return new Response();
        },
      })
      .delete('/')
      .send();
  });
  test('PATCH', async () => {
    anyClient
      .with({
        fetcher: async (url: string, init: RequestInit) => {
          expect(init?.method).toBe('PATCH');
          return new Response();
        },
      })
      .patch('/')
      .send();
  });
});

describe('Url', () => {
  const client = mockedClient;
  test('Can handle baseUrl with path', async () => {
    client
      .with({
        fetcher: async (url, init) => {
          expect(url).toBe(
            'https://petstore3.swagger.io/api/v3/pet/findByStatus'
          );
          return new Response();
        },
      })
      .get('/pet/findByStatus')
      .send();
  });
  test('Can override baseUrl', async () => {
    client
      .with({
        baseUrl: 'https://google.com/',
        fetcher: async (url, init) => {
          expect(url).toBe('https://google.com/pet/findByStatus');
          return new Response();
        },
      })
      .get('/pet/findByStatus')
      .send();
  });
  test('Replaces path params', async () => {
    client
      .with({
        fetcher: async (url, init) => {
          expect(url).toBe('https://petstore3.swagger.io/api/v3/pet/1');
          return new Response();
        },
      })
      .get('/pet/{petId}')
      .path({
        petId: 1,
      })
      .send();
  });
});

describe('Headers', () => {
  test('Can set default headers', async () => {
    const client = mockedClient.with({
      headers: {
        'x-default': 'header',
      },
    });
    client
      .with({
        fetcher: async (url, init) => {
          expect(init?.headers).toEqual({
            'x-default': 'header',
          });
          return new Response();
        },
      })
      .get('/pet/findByStatus')
      .send();
  });

  test('Can override default header', async () => {
    const client = mockedClient.with({
      headers: {
        api_key: '1234',
      },
    });

    client
      .with({
        fetcher: async (url, init) => {
          expect(init?.headers).toEqual({
            api_key: 'abcd',
          });
          return new Response();
        },
      })
      .delete('/pet/{petId}')
      .__path({})
      .headers({
        api_key: 'abcd',
      })
      .send();
  });

  test('Can add additional header per request', async () => {
    const client = mockedClient.with({
      headers: {
        'x-default': 'header',
      },
    });

    client
      .with({
        fetcher: async (url, init) => {
          expect(init?.headers).toEqual({
            'x-default': 'header',
            api_key: 'abcd',
          });
          return new Response();
        },
      })
      .delete('/pet/{petId}')
      .__path({})
      .headers({
        api_key: 'abcd',
      })
      .send();
  });
});

describe('Body', () => {
  const client = mockedClient;
  test('Using body serializes json', async () => {
    client
      .with({
        fetcher: async (url, init) => {
          const body = init?.body?.valueOf();
          expect(body).toBeTypeOf('string');
          expect(JSON.parse(body as string)).toEqual({
            name: 'hello',
          });
          return new Response();
        },
      })
      .post('/pet')
      .body({
        name: 'hello',
      } as any)
      .send();
  });
  test('Using body sets application/json header', async () => {
    client
      .with({
        fetcher: async (url, init) => {
          const headers = new Headers(init.headers);
          expect(headers.get('Content-Type')).toBe('application/json');
          return new Response();
        },
      })
      .post('/pet')
      .body({} as any)
      .send();
  });
});

describe('Form', () => {
  const client = mockedClient;
  test('Using form sets application/x-www-form-urlencoded header', async () => {
    client
      .with({
        fetcher: async (url, init) => {
          const headers = new Headers(init.headers);
          expect(headers.get('Content-Type')).toBe(
            'application/x-www-form-urlencoded'
          );
          return new Response();
        },
      })
      .post('/pet')
      .form({} as any)
      .send();
  });
});

describe('Middleware', () => {
  test('Can add middleware', async () => {
    const middleware = vi.fn();
    const client = mockedClient.with({
      middlewares: [middleware],
    });
    expect(middleware).toHaveBeenCalledTimes(0);
    await client.get('/pet/findByStatus').send();
    expect(middleware).toHaveBeenCalledTimes(1);
  });

  test('Can modify request with middleware', async () => {
    const middleware = vi.fn().mockImplementation((url, init, next) => {
      const body = 'bodyString';
      const headers = new Headers(init.headers);
      headers.set('x-middleware', 'yes');

      return next('https://google.com', {
        ...init,
        headers: headers,
        body: body,
      });
    });
    await mockedClient
      .with({
        middlewares: [middleware],
        fetcher: async (url, init) => {
          expect(init?.body).toBe('bodyString');
          expect(init?.headers).toEqual(new Headers({ 'x-middleware': 'yes' }));
          return new Response();
        },
      })
      .get('/pet/findByStatus')
      .send();
    expect(middleware).toHaveBeenCalledTimes(1);
  });

  test('Middleware are run in sequence', async () => {
    const firstMiddleware = vi.fn().mockImplementation((url, init, next) => {
      return next('https://first.com', init);
    });
    const secondMiddleware = vi.fn().mockImplementation((url, init, next) => {
      expect(url).toBe('https://first.com');
      return next('https://second.com', init);
    });
    await mockedClient
      .with({
        middlewares: [firstMiddleware, secondMiddleware],
        fetcher: async (url, init) => {
          expect(url).toBe('https://second.com');
          return new Response();
        },
      })
      .get('/pet/findByStatus')
      .send();
    expect(firstMiddleware).toHaveBeenCalledTimes(1);
    expect(secondMiddleware).toHaveBeenCalledTimes(1);
  });
});

describe('Retries', () => {
  const throwingFetcher: Fetcher = async (url, init) => {
    throw new Error('');
  };
  const fetcherWith500Response: Fetcher = async (url, init) => {
    return new Response(undefined, {
      status: 500,
    });
  };
  const no500StatusCondition: ConditionFunction = (response) => {
    return response.status < 500;
  };

  describe('Thrown requests', () => {
    test('Will retry if retries > 0', async () => {
      const mockedThrowingFetcher = vi.fn().mockImplementation(throwingFetcher);
      const client = mockedClient.with({
        fetcher: mockedThrowingFetcher,
        retries: 1,
      });
      expect(mockedThrowingFetcher).toHaveBeenCalledTimes(0);
      await client
        .get('/pet/findByStatus')
        .send()
        .catch(() => {});
      expect(mockedThrowingFetcher).toHaveBeenCalledTimes(2);
    });

    test('Wont retry if retries = 0', async () => {
      const mockedThrowingFetcher = vi.fn().mockImplementation(throwingFetcher);
      const client = mockedClient.with({
        fetcher: mockedThrowingFetcher,
        retries: 0,
      });
      expect(mockedThrowingFetcher).toHaveBeenCalledTimes(0);
      await client
        .get('/pet/findByStatus')
        .send()
        .catch(() => {});
      expect(mockedThrowingFetcher).toHaveBeenCalledTimes(1);
    });

    test('Wont retry if request aborted', async () => {
      const abortController = new AbortController();
      const mockedRealFetch = vi.fn().mockImplementation(fetch);
      const client = mockedClient.with({
        fetcher: mockedRealFetch,
        retries: 1,
      });
      abortController.abort();

      expect(mockedRealFetch).toHaveBeenCalledTimes(0);
      expect(() =>
        client
          .with({ baseUrl: 'https://this-domain-will-cause-dns-error.com' })
          .get('/pet/findByStatus')
          .send({ signal: abortController.signal })
      ).rejects.toThrowErrorMatchingSnapshot();
      expect(mockedRealFetch).toHaveBeenCalledTimes(1);
    });
  });

  describe('Failed condition', () => {
    test('Will retry if retries > 0', async () => {
      const mockedFetcherWith500Response = vi
        .fn()
        .mockImplementation(fetcherWith500Response);
      const client = mockedClient.with({
        fetcher: mockedFetcherWith500Response,
        retries: 1,
        additionalRetryCondition: no500StatusCondition,
      });
      expect(mockedFetcherWith500Response).toHaveBeenCalledTimes(0);
      await client.get('/pet/findByStatus').send();
      expect(mockedFetcherWith500Response).toHaveBeenCalledTimes(2);
    });

    test('Wont retry if retries = 0', async () => {
      const mockedFetcherWith500Response = vi
        .fn()
        .mockImplementation(fetcherWith500Response);
      const client = mockedClient.with({
        fetcher: mockedFetcherWith500Response,
        retries: 0,
        additionalRetryCondition: no500StatusCondition,
      });
      expect(mockedFetcherWith500Response).toHaveBeenCalledTimes(0);
      await client.get('/pet/findByStatus').send();
      expect(mockedFetcherWith500Response).toHaveBeenCalledTimes(1);
    });
  });

  describe('Passed condition', () => {
    test('Wont retry if retries > 0', async () => {
      const mockedFetcherWith200Response = vi
        .fn()
        .mockImplementation(fetcherWith200Response);
      const client = mockedClient.with({
        fetcher: mockedFetcherWith200Response,
        retries: 1,
        additionalRetryCondition: no500StatusCondition,
      });
      expect(mockedFetcherWith200Response).toHaveBeenCalledTimes(0);
      await client.get('/pet/findByStatus').send();
      expect(mockedFetcherWith200Response).toHaveBeenCalledTimes(1);
    });

    test('Wont retry if retries = 0', async () => {
      const mockedFetcherWith200Response = vi
        .fn()
        .mockImplementation(fetcherWith200Response);
      const client = mockedClient.with({
        fetcher: mockedFetcherWith200Response,
        retries: 0,
        additionalRetryCondition: no500StatusCondition,
      });
      expect(mockedFetcherWith200Response).toHaveBeenCalledTimes(0);
      await client.get('/pet/findByStatus').send();
      expect(mockedFetcherWith200Response).toHaveBeenCalledTimes(1);
    });
  });

  describe('Middlewares', () => {
    test('Called once for thrown retries', async () => {
      const middleware = vi.fn().mockImplementation((url, init, next) => {
        return next(url, init);
      });

      const client = mockedClient.with({
        middlewares: [middleware],
        fetcher: throwingFetcher,
        retries: 1,
      });

      expect(middleware).toHaveBeenCalledTimes(0);
      await client
        .get('/pet/findByStatus')
        .send()
        .catch(() => {});
      expect(middleware).toHaveBeenCalledTimes(1);
    });
    test('Called once for status retries', async () => {
      const middleware = vi.fn().mockImplementation((url, init, next) => {
        return next(url, init);
      });

      const client = mockedClient.with({
        middlewares: [middleware],
        fetcher: fetcherWith500Response,
        additionalRetryCondition: no500StatusCondition,
        retries: 1,
      });

      expect(middleware).toHaveBeenCalledTimes(0);
      await client
        .get('/pet/findByStatus')
        .send()
        .catch(() => {});
      expect(middleware).toHaveBeenCalledTimes(1);
    });
  });
});

describe('Formatters', () => {
  describe('Body', () => {
    test('Uses defaultFormatter by default', async () => {
      mockedClient
        .with({
          fetcher: async (url, init) => {
            const form = init?.body as URLSearchParams;
            expect(form.getAll('array')).toEqual(['1', '2']);

            return new Response();
          },
        })
        .post('/pet')
        .form({
          array: [1, 2],
        } as any)
        .send();
    });

    test('Can use joinFormatter', async () => {
      mockedClient
        .with({
          formFormatter: joinFormatter,
          fetcher: async (url, init) => {
            const form = init?.body as URLSearchParams;
            expect(form.get('array')).toBe('1,2');

            return new Response();
          },
        })
        .post('/pet')
        .form({ array: [1, 2] } as any)
        .send();
    });

    test('Can use pathFormatter', async () => {
      mockedClient
        .with({
          formFormatter: pathFormatter,
          fetcher: async (url, init) => {
            const form = init?.body as URLSearchParams;
            expect(form.get('array[0]')).toBe('1');
            expect(form.get('array[1]')).toBe('2');

            return new Response();
          },
        })
        .post('/pet')
        .form({ array: [1, 2] } as any)
        .send();
    });
  });

  describe('Query', () => {
    test('Uses defaultFormatter by default', async () => {
      mockedClient
        .with({
          fetcher: async (url, init) => {
            const searchParams = new URL(url).searchParams;
            expect(searchParams.getAll('tags')).toEqual(['1', '2']);

            return new Response();
          },
        })
        .get('/pet/findByTags')
        .query({
          tags: ['1', '2'],
        })
        .send();
    });

    test('Can use joinFormatter', async () => {
      mockedClient
        .with({
          formFormatter: joinFormatter,
          fetcher: async (url, init) => {
            const searchParams = new URL(url).searchParams;
            expect(searchParams.get('tags')).toBe('1,2');

            return new Response();
          },
        })
        .get('/pet/findByTags')
        .query({
          tags: ['1', '2'],
        })
        .send();
    });

    test('Can use pathFormatter', async () => {
      mockedClient
        .with({
          formFormatter: pathFormatter,
          fetcher: async (url, init) => {
            const searchParams = new URL(url).searchParams;
            expect(searchParams.get('tags[0]')).toBe('1');
            expect(searchParams.get('tags[1]')).toBe('2');

            return new Response();
          },
        })
        .get('/pet/findByTags')
        .query({
          tags: ['1', '2'],
        })
        .send();
    });
  });
});
