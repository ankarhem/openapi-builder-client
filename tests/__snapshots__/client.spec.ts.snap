// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Retries > Thrown requests > Wont retry if request aborted 1`] = `[AbortError: This operation was aborted]`;
